src/backend/distributed/utils/metadata_cache.c

```diff
@@ -745,7 +745,13 @@ BuildCachedShardList(DistTableCacheEntry *cacheEntry)
 
 		shardEntry = hash_search(DistShardCacheHash, &shardInterval->shardId, HASH_ENTER,
 								 &foundInCache);
-		Assert(!foundInCache);
+		if (foundInCache)
+		{
+			ereport(ERROR, (errmsg("a shard has changed the table it belongs to within "
+								   "this session"),
+							errhint("try again in a new session")));
```

Comment:
Hints are supposed to be actual sentences:
https://www.postgresql.org/docs/devel/static/error-style-guide.html
maybe something like "Reconnect and try again."

Also, I'd be a bit less precise about the error message, there could be other reasons.  How about "cached metadata for relation %u, shard %u is inconsistent"?